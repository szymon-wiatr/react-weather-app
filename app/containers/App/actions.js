import {
  LOAD_WEATHER,
  LOAD_WEATHER_SUCCESS,
  LOAD_WEATHER_ERROR,
} from './constants';

/**
 * Load the weather, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_WEATHER
 */
export function loadWeather() {
  return {
    type: LOAD_WEATHER,
  };
}

/**
 * Dispatched when the weather are loaded by the request saga
 *
 * @param  {array} weather The weather data
 * @param  {string} cityname The current cityname
 *
 * @return {object} An action object with a type of LOAD_WEATHER_SUCCESS passing the weather
 */
export function weatherLoaded(weather, cityname) {
  return {
    type: LOAD_WEATHER_SUCCESS,
    weather,
    cityname,
  };
}

/**
 * Dispatched when loading the wheater fails
 *
 * @param  {object} error The error
 *
 * @return {object} An action object with a type of LOAD_WEATHER_ERROR passing the error
 */
export function weatherLoadingError(error) {
  return {
    type: LOAD_WEATHER_ERROR,
    error,
  };
}
