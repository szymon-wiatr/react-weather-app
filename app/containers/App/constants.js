export const LOAD_WEATHER = 'weatherapp/App/LOAD_WEATHER';
export const LOAD_WEATHER_SUCCESS = 'weatherapp/App/LOAD_WEATHER_SUCCESS';
export const LOAD_WEATHER_ERROR = 'weatherapp/App/LOAD_WEATHER_ERROR';
