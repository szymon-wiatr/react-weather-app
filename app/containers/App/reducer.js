import produce from 'immer';
import {
  LOAD_WEATHER_SUCCESS,
  LOAD_WEATHER,
  LOAD_WEATHER_ERROR,
} from './constants';

export const initialState = {
  loading: false,
  error: false,
  currentCity: false,
  weatherData: {
    weather: false,
  },
};

/* eslint-disable default-case, no-param-reassign */
const appReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case LOAD_WEATHER:
        draft.loading = true;
        draft.error = false;
        draft.weatherData.weather = false;
        break;

      case LOAD_WEATHER_SUCCESS:
        [draft.weatherData.weather] = action.weather.data;
        draft.loading = false;
        draft.currentCity = action.cityname;
        break;

      case LOAD_WEATHER_ERROR:
        draft.error = action.error;
        draft.loading = false;
        break;
    }
  });

export default appReducer;
