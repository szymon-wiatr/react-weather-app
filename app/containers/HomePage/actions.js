import { CHANGE_CITYNAME } from './constants';

/**
 * Changes the input field of the form
 *
 * @param  {string} cityname The new text of the input field
 *
 * @return {object} An action object with a type of CHANGE_CITYNAME
 */
export function changeCityname(cityname) {
  return {
    type: CHANGE_CITYNAME,
    cityname,
  };
}
