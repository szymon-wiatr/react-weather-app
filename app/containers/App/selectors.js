import { createSelector } from 'reselect';
import { initialState } from './reducer';

const selectGlobal = state => state.global || initialState;

const makeSelectLoading = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.loading,
  );

const makeSelectError = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.error,
  );

const makeSelectWeather = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.weatherData.weather,
  );

export { selectGlobal, makeSelectLoading, makeSelectError, makeSelectWeather };
