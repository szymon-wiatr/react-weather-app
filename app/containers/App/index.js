import React from 'react';
import styled from 'styled-components';
import { Switch, Route } from 'react-router-dom';

import HomePage from 'containers/HomePage/Loadable';
import NotFoundPage from 'components/NotFoundPage/Loadable';

import GlobalStyle from '../../global-styles';

const AppWrapper = styled.div`
  max-width: calc(768px + 16px * 2);
  margin: 0 auto;
  padding: 16px;
`;

export default function App() {
  return (
    <AppWrapper>
      <Switch>
        <Route exact path="/" component={HomePage} />
        <Route path="" component={NotFoundPage} />
      </Switch>
      <GlobalStyle />
    </AppWrapper>
  );
}
