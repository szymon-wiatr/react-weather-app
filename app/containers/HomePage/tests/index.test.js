import React from 'react';
import { render } from 'react-testing-library';
import { Provider } from 'react-redux';
import { browserHistory } from 'react-router-dom';

import { HomePage, mapDispatchToProps } from '../index';
import { changeCityname } from '../actions';
import { loadWeather } from '../../App/actions';
import configureStore from '../../../configureStore';

describe('<HomePage />', () => {
  let store;

  beforeAll(() => {
    store = configureStore({}, browserHistory);
  });

  it('should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <HomePage loading={false} error={false} weather={{}} />
      </Provider>,
    );
    expect(firstChild).toMatchSnapshot();
  });

  it('should fetch the weather on mount if a cityname exists', () => {
    const submitSpy = jest.fn();
    render(
      <Provider store={store}>
        <HomePage
          cityname="Not Empty"
          onChangeCityname={() => {}}
          onSubmitForm={submitSpy}
        />
      </Provider>,
    );
    expect(submitSpy).toHaveBeenCalled();
  });

  it('should not call onSubmitForm if cityname is an empty string', () => {
    const submitSpy = jest.fn();
    render(
      <Provider store={store}>
        <HomePage onChangeCityname={() => {}} onSubmitForm={submitSpy} />
      </Provider>,
    );
    expect(submitSpy).not.toHaveBeenCalled();
  });

  it('should not call onSubmitForm if cityname is null', () => {
    const submitSpy = jest.fn();
    render(
      <Provider store={store}>
        <HomePage
          cityname=""
          onChangeCityname={() => {}}
          onSubmitForm={submitSpy}
        />
      </Provider>,
    );
    expect(submitSpy).not.toHaveBeenCalled();
  });

  describe('mapDispatchToProps', () => {
    describe('onChangeCityname', () => {
      it('should be injected', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        expect(result.onChangeCityname).toBeDefined();
      });

      it('should dispatch changeCityname when called', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        const cityname = 'SkarÅ¼ysko-Kamienna';
        result.onChangeCityname({ target: { value: cityname } });
        expect(dispatch).toHaveBeenCalledWith(changeCityname(cityname));
      });
    });

    describe('onSubmitForm', () => {
      it('should be injected', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        expect(result.onSubmitForm).toBeDefined();
      });

      it('should dispatch loadWeather when called', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        result.onSubmitForm();
        expect(dispatch).toHaveBeenCalledWith(loadWeather());
      });

      it('should preventDefault if called with event', () => {
        const preventDefault = jest.fn();
        const result = mapDispatchToProps(() => {});
        const evt = { preventDefault };
        result.onSubmitForm(evt);
        expect(preventDefault).toHaveBeenCalledWith();
      });
    });
  });
});
