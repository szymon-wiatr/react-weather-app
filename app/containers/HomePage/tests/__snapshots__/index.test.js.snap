// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> should render and match the snapshot 1`] = `
<section
  class="Section-c961xk-0 jdaQwv"
>
  <form
    class="Form-sc-1agrddf-0 chWgAW"
  >
    <input
      class="Input-fl4fxf-0 kIXWOE"
      id="cityname"
      placeholder="City"
      type="text"
      value=""
    />
  </form>
  <div
    class="Wrapper-sc-1ju4keu-0 jKiBZy"
  >
    <h1
      class="forecast-temperature"
    >
       â„ƒ
    </h1>
    <span
      aria-labelledby="sunrise"
      class="sun-icon"
      role="img"
    >
      ðŸŒ… 
    </span>
    <span
      aria-labelledby="sunset"
      class="sun-icon"
      role="img"
    >
      ðŸŒ„ 
    </span>
  </div>
</section>
`;
